Fix ec point formats

mattcaswell opened the pull request:
<!--
Thank you for your pull request. Please review below requirements.

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] tests are added or updated
- [x] CLA is signed

##### Description of change
<!-- Provide a description of the changes.

If it fixes a github issue, add Fixes #XXXX.
-->
Fix the EC point formats extension. This should be sent in the ServerHello if an EC based ciphersuite is
negotiated (<= TLS1.2). The relevant flag to do this was missed off in the recent extensions refactor. This bug should have been caught by `70-test_sslmessages` but that test never tries an EC ciphersuite, so we also update the test to do that.

Fixes #2133

richsalz commented on util/TLSProxy/Message.pm:
```diff
@@ -83,6 +83,10 @@ use constant {
 };
 
+use constant {
+    CIPHER_ADH_AES_128_SHA => 0x03000034
```
should this match the name in ssl.h?

mattcaswell replied:
None of the other constants in this file match the equivalents in the C. To me this is a fairly intuitive name (it more or less matches the OpenSSL "string" name of the ciphersuite (except "_" not "-" and a "CIPHER_" prefix)).

richsalz commented on util/TLSProxy/Proxy.pm:
```diff
@@ -535,4 +537,13 @@ sub reneg
 
+sub ciphersuite
+{
+    my $class = shift;
```
class or self?  is this line needed?

levitte replied:
class.  `$ciphersuite` is a class variable (i.e. a module global variable), not an instance variable.  It's the same as `is_tls13` about 10 lines up.

levitte commented on test/recipes/70-test_sslmessages.t:
```diff
@@ -46,6 +46,8 @@ my $proxy = TLSProxy::Proxy->new(
         checkhandshake::ALL_HANDSHAKES
         & ~checkhandshake::RESUME_HANDSHAKE],
+    [TLSProxy::Message::MT_SERVER_KEY_EXCHANGE,
+        checkhandshake::EC_HANDSHAKE],
```
You should wrap this with a `diabled("ec")` check...

mattcaswell replied:
In #2154 I disable the whole test for no-ec. I think its better than adding lots of complexity to the test to deal with this option.

levitte replied:
It's not a whole lot of complexity, here's how you do it:

``` perl
    (disabled("ec") ? () :
     [TLSProxy::Message::MT_SERVER_KEY_EXCHANGE,
         checkhandshake::EC_HANDSHAKE]),
```

levitte commented on test/recipes/70-test_sslmessages.t:
```diff
@@ -135,6 +137,8 @@ my $proxy = TLSProxy::Proxy->new(
     [TLSProxy::Message::MT_SERVER_HELLO, TLSProxy::Message::EXT_NPN,
         checkhandshake::NPN_SRV_EXTENSION],
+    [TLSProxy::Message::MT_SERVER_HELLO, TLSProxy::Message::EXT_EC_POINT_FORMATS,
+        checkhandshake::EC_POINT_FORMAT_SRV_EXTENSION],
```
... and this too

mattcaswell replied:
As above.

mattcaswell replied:
This isn't actually necessary because we skip the test that uses this information anyway.

levitte replied:
Okie

levitte commented on test/recipes/70-test_sslmessages.t:
```diff
@@ -358,3 +362,13 @@ checkhandshake($proxy, checkhandshake::DEFAULT_HANDSHAKE,
+checkhandshake($proxy, checkhandshake::EC_HANDSHAKE,
+               checkhandshake::DEFAULT_EXTENSIONS
+               | checkhandshake::EC_POINT_FORMAT_SRV_EXTENSION,
+               "EC handshake test");
```
Wrap this with:

``` perl
SKIP: {
    skip "No EC support in this OpenSSL build", 1 if disabled("ec");

    # original code here

}
```

mattcaswell replied:
As above.

mattcaswell replied:
Done

FdaSilvaYY commented:
FYI, the `no-ec` option is broken.
Some missing #if guards in the new extension-handling code.

mattcaswell commented:
> FYI, the no-ec option is broken.
Some missing #if guards in the new extension-handling code.

Yeah, I know. I'm working on it.

mattcaswell commented:
Fixed issues above. New commits added taken from #2154. Also added a new commit to deal with a hang in test_sslmessages when using no-ec. This is because the ct tests in that file require ec to be available, so I just disabled those in the event of no-ec.

levitte commented on test/recipes/70-test_sslmessages.t:
```diff
@@ -94,10 +97,14 @@ my $proxy = TLSProxy::Proxy->new(
+    (disabled("ec") ? () :
+                      [TLSProxy::Message::MT_CLIENT_HELLO,
+                       TLSProxy::Message::EXT_EC_POINT_FORMATS,
+                       checkhandshake::DEFAULT_EXTENSIONS]),
```
Side comment:  that can be written like this as well:

``` perl
    (disabled("ec") ? () :
                      ([TLSProxy::Message::MT_CLIENT_HELLO,
                        TLSProxy::Message::EXT_SUPPORTED_GROUPS,
                        checkhandshake::DEFAULT_EXTENSIONS],
                       [TLSProxy::Message::MT_CLIENT_HELLO,
                        TLSProxy::Message::EXT_EC_POINT_FORMATS,
                        checkhandshake::DEFAULT_EXTENSIONS])),
```

This is because lists of this kind are flattened in Perl.  Also, I'm not 100% sure the outer parens are strictly needed.

But, this is a preference and your choice to make.

mattcaswell replied:
Actually, I did think about doing it that way, but made a deliberate decision not to. Either way is valid, but supported_groups may include non ec groups in the not too distant future if we implement support for DHE groups, so I decided to have separate expressions for them.

levitte replied:
Cool.

levitte commented on test/recipes/70-test_sslmessages.t:
```diff
@@ -358,3 +369,16 @@ checkhandshake($proxy, checkhandshake::DEFAULT_HANDSHAKE,
+SKIP: {
+    skip "No EC support in this OpenSSL build", 1 if disabled("ec");
+    $proxy->clear();
+    $proxy->clientflags("-no_tls1_3");
```
Is that flag strictly necessary?  I would prefer if it's avoided as much as possible.  My experience with sslvertol has shown that it's not as necessary as it was once...

mattcaswell replied:
Yes, it is very much necessary. Without it TLSv1.3 would be negotiated and then the connection would fail due to a "no shared cipher". WRT sslvertol, I think you have a misunderstanding and I will comment on the relevant PR.

levitte replied:
Okie.

mattcaswell commented:
Pushed. Thanks.
